boring/src/ssl/mod.rs

```diff
@@ -1560,6 +1629,33 @@ pub struct CipherBits {
     pub algorithm: i32,
 }
 
+#[repr(transparent)]
+pub struct ClientHello(ffi::SSL_CLIENT_HELLO);
+
+impl ClientHello {
+    /// Returns the data of a given extension, if present.
+    ///
+    /// This corresponds to [`SSL_early_callback_ctx_extension_get`].
+    ///
+    /// [`SSL_early_callback_ctx_extension_get`]: https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_early_callback_ctx_extension_get
+    pub fn get_extension(&self, type_: ExtensionType) -> Option<&[u8]> {
+        let slice = unsafe {
+            let mut ptr = ptr::null();
+            let mut len = 0;
+            let result =
+                ffi::SSL_early_callback_ctx_extension_get(&self.0, type_.0, &mut ptr, &mut len);
+            if result == 0 {
+                return None;
+            }
+            slice::from_raw_parts(ptr, len)
+        };
+        let (len, rest) = slice.split_at(2);
+        let len = (len[0] as usize) << 8 | len[1] as usize;
+        assert_eq!(len, rest.len());
```

Comment:
Not sure if we want to panic here, considering we're working with the user-provided data. We should check for the length prefix to be exactly 2 bytes and that this invariant holds and return `Err(MalformedExtension)` or smthg, if it's not

Reply:
Thinking twice about it I feel like this is dependent on the expected type of the extension data of each extension and it should probably not be trimmed here, I'll check the various specs to confirm.

Reply:
First, [RFC 5246](https://tools.ietf.org/html/rfc5246#section-4.3) says:

> A vector (single-dimensioned array) is a stream of homogeneous data elements.  The size of the vector may be specified at documentation time or left unspecified until runtime. In either case, the length declares the number of bytes, not the number of elements, in the vector.

Then it says:

> Variable-length vectors are defined by specifying a subrange of legal lengths, inclusively, using the notation `<floor..ceiling>`.  When these are encoded, the actual length precedes the vector's contents in the byte stream.  The length will be in the form of a number consuming as many bytes as required to hold the vector's specified maximum (`ceiling`) length. A variable-length vector with an actual length field of zero is referred to as an empty vector.

So I didn't know that the lengths of arrays in those RFCs are in terms of *bytes* not in terms of the number of elements. This is why I assumed that the length we are now stripping was the length of the opaque payload of the extension. I can now see that this is wrong.

Furthermore, [RFC 7301](https://tools.ietf.org/html/rfc7301#section-3.1) says:

>    The "extension_data" field of the ("application_layer_protocol_negotiation(16)") extension SHALL contain a "ProtocolNameList" value.
> 
> ```
> opaque ProtocolName<1..2^8-1>;
> 
> struct {
>     ProtocolName protocol_name_list<2..2^16-1>
> } ProtocolNameList;
> ```

So that length is indeed always 2 bytes in the case of the ALPN extension, and we shouldn't be stripping it as this is not an ALPN-specific method.

Reply:
I have reverted the trimming as this is not the place where we should be doing that.